File: Data Structures & Algorithms/combination-target-sum/submission-0.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        dfs(nums, 0, curr, res, target);
        return res;
    }

    private void dfs(int[] nums, int i, List<Integer> curr, List<List<Integer>> res, int target) {
        if (target == 0) {
            res.add(new ArrayList<>(curr));
            return;
        }
        if (i >= nums.length || target < 0) {
            return;
        }

        curr.add(nums[i]);
        dfs(nums, i, curr, res, target - nums[i]);
        curr.remove(curr.size() - 1);
        dfs(nums, i + 1, curr, res, target);
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-0.py
class Solution:
    def dfs(self, nums, i, subset, res):
        if i >= len(nums): 
            res.append(subset.copy())
            return 

        subset.append(nums[i])
        self.dfs(nums, i + 1, subset, res);
        subset.pop();
        while (i + 1 < len(nums) and nums[i] == nums[i + 1]):
            i += 1
        self.dfs(nums, i + 1, subset, res);

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        subset = []
        nums.sort()
        self.dfs(nums, 0, subset, res)
        return res